mautic/mautic
# Remove sub query mechanism from segment membership filter

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) 
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 




#### Description:
This is our way how to handle large instance and segment membership filter usage.
Mautic generate query inside query. That means If you use segment membership then Mautic build query and inside query build query from parent segment. This make extra large SQL quiery for segment with more filters and combination with segment membership filter on large datasets

After PR query just check `lead_lists_leads` membership + try rebuild simple segments (without segment membership ) first. This save use a lot of time and lot of CPU/RAM resources

Segment gmail 

![image](https://user-images.githubusercontent.com/462477/150351524-f7a8ac1c-74b6-4d11-b65b-4fbc37eb4f8d.png)

Segment hotmail

![image](https://user-images.githubusercontent.com/462477/150351609-09c73323-b9ad-4b99-ae18-45dc5d64c960.png)

Segment complex

![image](https://user-images.githubusercontent.com/462477/150351693-40cbeebb-c841-4b7d-a65a-370d7931c98a.png)


Before query looks like

`SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT orp.lead_id as leadIdPrimary, orp.lead_id as id, orp.leadlist_id FROM lead_lists_leads orp LEFT JOIN (SELECT l.id, parz.id as parz_id, par1a.id as par1a_id FROM leads l LEFT JOIN (SELECT l.id FROM leads l LEFT JOIN lead_lists_leads parx ON l.id = parx.lead_id and parx.leadlist_id = 10 and (parx.manually_removed = 1) WHERE (((l.email IS NOT NULL) AND (l.email <> '')) AND (l.generated_email_domain = 'gmail') AND ((l.firstname IS NOT NULL) AND (l.firstname <> '')) AND ((l.lastname IS NOT NULL) AND (l.lastname <> '')) AND (parx.lead_id IS NULL)) OR (l.id IN (SELECT pary.lead_id FROM lead_lists_leads pary WHERE (pary.leadlist_id = 10) AND ((pary.manually_added = 1) OR (pary.manually_removed = '')))))  parz ON parz.id = l.id LEFT JOIN (SELECT l.id FROM leads l LEFT JOIN lead_lists_leads par18 ON l.id = par18.lead_id and par18.leadlist_id = 11 and (par18.manually_removed = 1) WHERE (((l.email IS NOT NULL) AND (l.email <> '')) AND (l.generated_email_domain = 'hotmail') AND ((l.firstname IS NOT NULL) AND (l.firstname <> '')) AND ((l.lastname IS NOT NULL) AND (l.lastname <> '')) AND (par18.lead_id IS NULL)) OR (l.id IN (SELECT par19.lead_id FROM lead_lists_leads par19 WHERE (par19.leadlist_id = 11) AND ((par19.manually_added = 1) OR (par19.manually_removed = '')))))  par1a ON par1a.id = l.id WHERE (parz.id IS NOT NULL) OR (par1a.id IS NOT NULL)) members ON members.id=orp.lead_id WHERE (orp.leadlist_id = 12) AND (members.id IS NULL) AND (orp.manually_added = '0')) sss`

After  query looks like

`SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT orp.lead_id as leadIdPrimary, orp.lead_id as id, orp.leadlist_id FROM lead_lists_leads orp LEFT JOIN (SELECT l.id, par8.id as par8_id, para.id as para_id FROM leads l LEFT JOIN (SELECT l.id FROM leads l WHERE l.id IN (SELECT par7.lead_id FROM lead_lists_leads par7 WHERE (par7.leadlist_id = 10) AND ((par7.manually_added = 1) OR (par7.manually_removed = ''))))  par8 ON par8.id = l.id LEFT JOIN (SELECT l.id FROM leads l WHERE l.id IN (SELECT par9.lead_id FROM lead_lists_leads par9 WHERE (par9.leadlist_id = 11) AND ((par9.manually_added = 1) OR (par9.manually_removed = ''))))  para ON para.id = l.id WHERE (par8.id IS NOT NULL) OR (para.id IS NOT NULL)) members ON members.id=orp.lead_id WHERE (orp.leadlist_id = 12) AND (members.id IS NULL) AND (orp.manually_added = '0'))`



#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2.  Create segment membership filters for segments
3. Try If `mautic:segments:rebuild` works like before. On large table you see extreme improvement




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10063"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
